File: LC/Medium/1642.py
class Solution:
    def furthestBuilding(self, heights: List[int], bricks: int, ladders: int) -> int:
        heap = []
        
        for i in range(len(heights)-1):
            dist = heights[i+1] - heights[i]

            if dist <= 0:
                continue
            
            heappush(heap, dist)

            if len(heap) <= ladders:
                continue
            
            bricks -= heappop(heap)
            
            if bricks < 0:
                return i
            
        return len(heights) - 1

File: LC/C++/Medium/1642.cpp
class Solution {
public:
    int furthestBuilding(vector<int>& heights, int bricks, int ladders) {
        int n = heights.size();
        std::priority_queue<int, std::vector<int>, std::greater<int>> pq;

        for (size_t i{}; i < n - 1; ++i) {
            int gap = heights[i + 1] - heights[i];
            if (gap <= 0) continue;

            pq.push(gap);
            if (pq.size() <= ladders) continue;

            bricks -= pq.top();
            pq.pop();

            if (bricks < 0) return i;
        }

        return n - 1;
    }
};